components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1269,7 +1270,139 @@ where
             SignificantMsg::RaftlogFetched { context, res } => {
                 self.on_raft_log_fetched(context, res);
             }
+            SignificantMsg::EnterForceLeaderState { failed_stores } => {
+                self.on_enter_pre_force_leader(failed_stores)
+            }
+            SignificantMsg::ExitForceLeaderState => self.on_exit_force_leader(),
+        }
+    }
+
+    fn on_enter_pre_force_leader(&mut self, failed_stores: HashSet<u64>) {
+        if self.fsm.peer.force_leader.is_some() {
+            return;
+        }
+
+        let region = self.region();
+        let expected_alive_voter = self
+            .fsm
+            .peer
+            .raft_group
+            .raft
+            .prs()
+            .conf()
+            .voters()
+            .ids()
+            .iter()
+            .filter(|peer_id| {
+                let store_id = region
+                    .get_peers()
+                    .iter()
+                    .find(|p| p.get_id() == *peer_id)
+                    .unwrap()
+                    .get_store_id();
+                !failed_stores.contains(&store_id)
+            })
+            .collect();
+        info!(
+            "enter pre force leader state";
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+            "alive_voter" => ?expected_alive_voter,
+        );
+
+        if self.fsm.peer.is_leader() {
+            warn!(
+                "pre force leader on leader";
+                "region_id" => self.fsm.region_id(),
+                "peer_id" => self.fsm.peer_id(),
+            );
+            return;
+        }
+
+        // become candidate first to increase term
+        self.fsm.peer.raft_group.raft.become_candidate();
+
+        // trigger vote request to all voters, will check the vote result in `check_force_leader`
+        self.fsm.peer.raft_group.campaign().unwrap();
+        assert!(
+            self.fsm.peer.raft_group.raft.state == StateRole::PreCandidate
+                || self.fsm.peer.raft_group.raft.state == StateRole::Candidate
+        );
+        if !self
```

Comment:
How can it be possible?

Reply:
when there is a pending configuration change

Reply:
How can it be possible giving that wait apply is called before force leader?

Reply:
just defensive programming, even though it's impossible in normal process